unittests/finality_proof.hpp

```diff
@@ -10,25 +10,64 @@ namespace finality_proof {
-      uint32_t last_proposed_finalizer_policy_generation = 0;
-      digest_type action_mroot; //this is the real action_mroot, as returned from finality_data
-      digest_type base_digest;
-      digest_type active_finalizer_policy_digest;
-      digest_type last_pending_finalizer_policy_digest;
-      block_timestamp_type last_pending_finalizer_policy_start_timestamp;
-      digest_type last_proposed_finalizer_policy_digest;
-      digest_type finality_digest;
-      digest_type level_3_commitments_digest;
-      digest_type level_2_commitments_digest;
-      digest_type finality_leaf;
-      digest_type finality_root;
-      block_timestamp_type parent_timestamp;
+      signed_block_ptr block = nullptr;
+      qc_data_t qc_data{};
+      action_trace onblock_trace{};
+      finality_data_t finality_data{};
+      finalizer_policy active_finalizer_policy{};
+      finalizer_policy last_pending_finalizer_policy{};
+      finalizer_policy last_proposed_finalizer_policy{};
+      digest_type action_mroot{}; //this is the real action_mroot, as returned from finality_data
+      digest_type base_digest{};
+      block_timestamp_type last_pending_finalizer_policy_start_timestamp{};
+      digest_type finality_digest{};
+      level_3_commitments_t level_3_commitments{};
+      level_2_commitments_t level_2_commitments{};
+      digest_type finality_leaf{};
+      digest_type finality_root{};
+      block_timestamp_type parent_timestamp{};
+
+      digest_type level_3_commitments_digest() const {
+         return fc::sha256::hash(level_3_commitments);
+      }
+
+      digest_type level_2_commitments_digest() const {
+         return fc::sha256::hash(level_2_commitments);
+      }
+      
+   };
+
+   inline std::string bitset_to_input_string(const boost::dynamic_bitset<unsigned char>& bitset) {
+      static const char* hexchar = "0123456789abcdef";
+
+      boost::dynamic_bitset<unsigned char> bs(bitset);
+      bs.resize((bs.size() + 7) & ~0x7);
+      assert(bs.size() % 8 == 0);
+
+      std::string result;
+      result.resize(bs.size() / 4);
+      for (size_t i = 0; i < bs.size(); i += 4) {
+         size_t x = 0;
+         for (size_t j = 0; j < 4; ++j)
+            x += bs[i+j] << j;
+         auto slot = i / 4;
+         result[slot % 2 ? slot - 1 : slot + 1] = hexchar[x]; // flip the two hex digits for each byte
+      }
+      return result;
+   }
+
+   inline std::string binary_to_hex(const std::string& bin) {
+      boost::dynamic_bitset<unsigned char> bitset(bin.size());
+      for (size_t i = 0; i < bin.size(); ++i) {
+          if (bin[i] == '1') {
+              bitset.set(bin.size() - 1 - i);
+          }
+      }
+      return bitset_to_input_string(bitset);
+   }
+
+   inline auto finalizers_string = [](const vote_bitset_t finalizers)  {
+      return bitset_to_input_string(finalizers);
    };
 
    static digest_type hash_pair(const digest_type& a, const digest_type& b) {

```

Comment:
All of these `static` that are at `namespace` scope should be `inline`.

Reply:
Done